libraries/AP_Soaring/AP_Soaring.cpp

```diff
@@ -0,0 +1,381 @@
+    
+    // @Param: POLAR_K
+    // @DisplayName: Cl factor
+    // @Description: Cl factor 2*m*g/(rho*S)
+    // @Units: m*m/s/s
+    // @Range: 0 0.5
+    // @User: Advanced 
+    AP_GROUPINFO("POLAR_K", 11, SoaringController, polar_K, 25.6),
+    
+    // @Param: ALT_MAX
+    // @DisplayName: Maximum soaring altitude, relative to the home location
+    // @Description: Don't thermal any higher than this.
+    // @Units: meters
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_MAX", 12, SoaringController, alt_max, 350.0),
+    
+    // @Param: ALT_MIN
+    // @DisplayName: Minimum soaring altitude, relative to the home location
+    // @Description: Don't get any lower than this.
+    // @Units: meters
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_MIN", 13, SoaringController, alt_min, 50.0),
+    
+    // @Param: ALT_CUTOFF
+    // @DisplayName: Maximum power altitude, relative to the home location
+    // @Description: Cut off throttle at this alt.
+    // @Units: meters
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_CUTOFF", 14, SoaringController, alt_cutoff, 250.0),
+
+    AP_GROUPEND
+};
+
+void SoaringController::get_target(Location &wp)
+{
+    wp = _prev_update_location;
+    location_offset(wp, _ekf.X[2], _ekf.X[3]);
+}
+
+bool SoaringController::suppress_throttle()
+{
+    float alt = 0;
+    get_altitude_wrt_home(&alt);
+    
+    if (_throttle_suppressed && (alt < alt_min)) {
+        // Time to throttle up
+        _throttle_suppressed = false;
+    }
+    else if ((!_throttle_suppressed) && (alt > alt_cutoff)) {
+        // Start glide
+        _throttle_suppressed = true;
+        // Zero the pitch integrator - the nose is currently raised to climb, we need to go back to glide.
+        _spdHgt->reset_pitch_I();
+        _cruise_start_time_us = AP_HAL::micros64();
+        // Reset the filtered vario rate - it is currently elevated due to the climb rate and would otherwise take a while to fall again,
+        // leading to false positives.
+        _filtered_vario_reading = 0;
+    }
+          
+    return _throttle_suppressed;
+}
+
+bool SoaringController::check_thermal_criteria()
+{
+    return (soar_active 
+            && ((AP_HAL::micros64() - _cruise_start_time_us) > ((unsigned)min_cruise_s * 1e6))
+            && _filtered_vario_reading > thermal_vspeed 
+            && _alt < alt_max 
+            && _alt > alt_min);
+}
+
+bool SoaringController::check_cruise_criteria()
+{
+    float thermalability = (_ekf.X[0]*expf(-powf(_loiter_rad / _ekf.X[1], 2))) - EXPECTED_THERMALLING_SINK; 
+    
+    if (soar_active && (AP_HAL::micros64() - _thermal_start_time_us) > ((unsigned)min_thermal_s * 1e6) && thermalability < McCready(_alt)) {
+        hal.console->printf("Thermal weak, recommend quitting: W %f R %f th %f alt %f Mc %f\n", (double)_ekf.X[0], (double)_ekf.X[1], (double)thermalability, (double)_alt, (double)McCready(_alt));
```

Comment:
please don't use hal.console->printf() in flight, use GCS_MAVLink::send_statustext_all() instead. That way it appears as a proper MAVLink message, is logged correctly, goes to all GCS instances and doesn't much up the MAVLink serial stream.